src/smt/seq_unicode.cpp

```diff
@@ -67,67 +72,150 @@ namespace smt {
+                    propagate(assign_le(v, v_maxchar, null_literal));
+                    if (ctx().inconsistent()) return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    bool seq_unicode::enforce_char_codes(svector<theory_var> char_vars) {
+
+        // Iterate over all theory variables until the context is inconsistent
+        bool success = true;
         arith_util a(m);
         arith_value avalue(m);
         avalue.init(&ctx());
         uint_set seen;
-        for (unsigned v = 0; !ctx().inconsistent() && v < th.get_num_vars(); ++v) {
-            expr* e = th.get_expr(v);
-            if (!seq.is_char(e)) 
-                continue;
-            dl.init_var(v);            
-
-            int val = dl.get_assignment(v).get_int();
-            if (val > static_cast<int>(zstring::max_char())) {
-                expr_ref ch(seq.str.mk_char(zstring::max_char()), m);
-                enode* n = th.ensure_enode(ch);
-                theory_var v_max = n->get_th_var(th.get_id());
-                assign_le(v, v_max, null_literal);
-                added_constraint = true;
-                continue;
-            }
-            if (val < 0) {
-                theory_var v_min = ensure0();
-                assign_le(v_min, v, null_literal);
-                added_constraint = true;
-                continue;
-            }
-            if (seen.contains(val))
-                continue;
+        for (auto v : char_vars) {
+            if (ctx().inconsistent()) break;
+
+            // Make sure we haven't seen this value already
+            int val = get_value(v);
+            if (seen.contains(val)) continue;
             seen.insert(val);
-            // ensure str.to_code(unit(v)) = val
-            expr_ref ch(seq.str.mk_unit(seq.str.mk_char(val)), m);          
+
+            // Ensure str.to_code(unit(v)) = val
+            expr_ref ch(seq.str.mk_unit(seq.str.mk_char(val)), m);
             expr_ref code(seq.str.mk_to_code(ch), m);
             rational val2;
-            if (avalue.get_value(code, val2) && val2 == rational(val))
-                continue;
+
+            if (avalue.get_value(code, val2) && val2 == rational(val)) continue;
+
             add_axiom(th.mk_eq(code, a.mk_int(val), false));
-            added_constraint = true;
+            success = false;
         }
-        if (added_constraint)
-            return false;
-        
-        // ensure equalities over shared symbols
-        if (th.assume_eqs(m_var_value_table))
-            return false;
-        
+
+        // If a constraint was added without being propagated, we can't be finished yet
+        return success;
+    }
+
+    bool seq_unicode::final_check() {
+
+        // Get character variables
+        svector<theory_var> char_vars;
+        for (unsigned v = 0; v < th.get_num_vars(); ++v) {
+            if(seq.is_char(th.get_expr(v))) char_vars.push_back(v);
```

Comment:
we can prune this vector further using the filter:

```
th.get_enode(v)->is_root()
```
Only root nodes (nodes that are equivalence class representatives)
will be required to check. Equality enforcement should take care that all other nodes in the same equivalence class received the same value as the root.